swaybg/main.c

```diff
@@ -52,163 +68,264 @@ bool is_valid_color(const char *color) {
+	}
+}
+
+static void handle_global_remove(void *data, struct wl_registry *registry,
+		uint32_t name) {
+	// who cares
+}
 
-		cairo_surface_destroy(image);
+static const struct wl_registry_listener registry_listener = {
+	.global = handle_global,
+	.global_remove = handle_global_remove,
+};
+
+int main(int argc, const char **argv) {
+	struct swaybg_args args = {0};
+	struct swaybg_state state = {0};
+	state.args = &args;
+	wlr_log_init(L_DEBUG, NULL);
+
+	if (argc != 4) {
+		wlr_log(L_ERROR, "Do not run this program manually. "
+				"See man 5 sway and look for output options.");
+		return 1;
+	}
+	args.output_idx = atoi(argv[1]);
+	args.path = argv[2];
+	args.mode = atoi(argv[3]);
+
+	args.mode = BACKGROUND_MODE_STRETCH;
+	if (strcmp(argv[3], "stretch") == 0) {
+		args.mode = BACKGROUND_MODE_STRETCH;
+	} else if (strcmp(argv[3], "fill") == 0) {
+		args.mode = BACKGROUND_MODE_FILL;
+	} else if (strcmp(argv[3], "fit") == 0) {
+		args.mode = BACKGROUND_MODE_FIT;
+	} else if (strcmp(argv[3], "center") == 0) {
+		args.mode = BACKGROUND_MODE_CENTER;
+	} else if (strcmp(argv[3], "tile") == 0) {
+		args.mode = BACKGROUND_MODE_TILE;
+	} else if (strcmp(argv[3], "solid_color") == 0) {
+		args.mode = BACKGROUND_MODE_SOLID_COLOR;
+	} else {
+		wlr_log(L_ERROR, "Unsupported background mode: %s", argv[3]);
+		return 1;
+	}
+
+	if (!prepare_context(&state)) {
+		return 1;
 	}
 
-	while (wl_display_dispatch(registry->display) != -1);
+	state.display = wl_display_connect(NULL);
+	if (!state.display) {
+		wlr_log(L_ERROR, "Failed to create display\n");
+		return 1;
+	}
+
+	struct wl_registry *registry = wl_display_get_registry(state.display);
+	wl_registry_add_listener(registry, &registry_listener, &state);
+	wl_display_roundtrip(state.display);
+
+	if (!state.compositor) {
+		wlr_log(L_DEBUG, "wl-compositor not available");
+		return 1;
+	}
+	if (!state.layer_shell) {
+		wlr_log(L_ERROR, "layer-shell not available");
+		return 1;
+	}
+
+	state.surface = wl_compositor_create_surface(state.compositor);
+	if (!state.surface) {
+		wlr_log(L_ERROR, "failed to create wl_surface");
+		return 1;
+	}
+
+	state.layer_surface = zwlr_layer_shell_v1_get_layer_surface(
+			state.layer_shell, state.surface, state.output,
+			ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND, "wallpaper");
+	if (!state.layer_surface) {
+		wlr_log(L_ERROR, "failed to create zwlr_layer_surface");
+		return 1;
+	}
+	zwlr_layer_surface_v1_set_size(state.layer_surface, 0, 0);
+	zwlr_layer_surface_v1_set_anchor(state.layer_surface,
+			ZWLR_LAYER_SURFACE_V1_ANCHOR_TOP |
+			ZWLR_LAYER_SURFACE_V1_ANCHOR_RIGHT |
+			ZWLR_LAYER_SURFACE_V1_ANCHOR_BOTTOM |
+			ZWLR_LAYER_SURFACE_V1_ANCHOR_LEFT);
+	zwlr_layer_surface_v1_add_listener(state.layer_surface,
+			&layer_surface_listener, &state);
+	wl_surface_commit(state.surface);
+	wl_display_roundtrip(state.display);
 
-	for (i = 0; i < surfaces->length; ++i) {
-		struct window *window = surfaces->items[i];
-		window_teardown(window);
+	state.run_display = true;
```

Comment:
We might want to move this before the first commit in case the compositor closes the layer surface immediately.